include/nbl/builtin/hlsl/member_test_macros.hlsl

```diff
@@ -0,0 +1,53 @@
+// Copyright (C) 2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_MEMBER_TEST_MACROS_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MEMBER_TEST_MACROS_INCLUDED_
+
+#include <nbl/builtin/hlsl/type_traits.hlsl>
+
+#ifdef __HLSL_VERSION
+
+namespace nbl
+{
+namespace hlsl
+{
+
+namespace impl
+{
+
+template<class T>
+struct valid_expression : true_type {};
+
+template<> struct valid_expression<void> : false_type {};
```

Comment:
this is just `make_void` all over again?

Reply:
No it's different.  
`make_void` maps everything to void. 
`valid_expression` maps void to false and everything else to true
Rather than doing
```cpp
template<class T,class=void>
struct spec{ };
template<class T>
struct spec<T, typename enable_if<!is_same<decltype(expr), void>::value, void>::type>
```
we can do
```cpp
template<class T,class=bool_constant<true>>
struct spec{ };
template<class T>
struct spec<T, bool_constant<valid_expression<decltype(expr)>::value>> {};
```

Reply:
do you even need an extra `bool_constant` at that stage? wonder if:
```cpp
template<class T,bool=true>
struct spec{ };
template<class T>
struct spec<T,valid_expression<decltype(expr)>::value> {};
```
would just work?

Reply:
it doesnt Dxc complains about partial spec being type dependent